File: admin-app/src/stores/stats.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '@/services/api'

export interface UserStats {
  total_users: number
  growth_percentage: number
  active_users: number
  new_users_today: number
  new_users_week: number
}

export interface SessionStats {
  active_sessions: number
  change_percentage: number
  total_sessions_today: number
  unique_users_today: number
}

export interface OrgStats {
  total_organizations: number
  growth_percentage: number
  active_organizations: number
}

export interface ClientStats {
  total_clients: number
  change_percentage: number
  active_clients: number
}

export interface ActivityData {
  date: string
  logins: number
  registrations: number
  active_users: number
}

export interface LoginDistribution {
  org: string
  count: number
  percentage: number
}

export interface RecentActivity {
  id: string
  type: 'login' | 'registration' | 'user_update' | 'client_create'
  user_email: string
  org: string
  timestamp: string
  details?: string
}

export const useStatsStore = defineStore('stats', () => {
  const userStats = ref<UserStats | null>(null)
  const sessionStats = ref<SessionStats | null>(null)
  const orgStats = ref<OrgStats | null>(null)
  const clientStats = ref<ClientStats | null>(null)
  const activityData = ref<ActivityData[]>([])
  const loginDistribution = ref<LoginDistribution[]>([])
  const recentActivities = ref<RecentActivity[]>([])

  const lastLoginTime = ref<string | null>(null)
  const failedLoginsToday = ref<number>(0)
  const newUsersThisWeek = ref<number>(0)

  const loadUserStats = async (): Promise<void> => {
    try {
      const response = await api.get('/stats/users')
      userStats.value = response.data
    } catch (error) {
      console.error('Failed to load user stats:', error)
    }
  }

  const loadSessionStats = async (): Promise<void> => {
    try {
      const response = await api.get('/stats/sessions')
      sessionStats.value = response.data
    } catch (error) {
      console.error('Failed to load session stats:', error)
    }
  }

  const loadOrgStats = async (): Promise<void> => {
    try {
      const response = await api.get('/stats/organizations')
      orgStats.value = response.data
    } catch (error) {
      console.error('Failed to load organization stats:', error)
    }
  }

  const loadClientStats = async (): Promise<void> => {
    try {
      const response = await api.get('/stats/clients')
      clientStats.value = response.data
    } catch (error) {
      console.error('Failed to load client stats:', error)
    }
  }

  const loadActivityData = async (timeRange: string): Promise<void> => {
    try {
      const response = await api.get('/stats/activity', {
        params: { range: timeRange }
      })
      activityData.value = response.data
    } catch (error) {
      console.error('Failed to load activity data:', error)
    }
  }

  const loadLoginDistribution = async (timeRange: string): Promise<void> => {
    try {
      const response = await api.get('/stats/login-distribution', {
        params: { range: timeRange }
      })
      loginDistribution.value = response.data
    } catch (error) {
      console.error('Failed to load login distribution:', error)
    }
  }

  const loadRecentActivities = async (limit: number = 10): Promise<void> => {
    try {
      const response = await api.get('/stats/recent-activities', {
        params: { limit }
      })
      recentActivities.value = response.data
    } catch (error) {
      console.error('Failed to load recent activities:', error)
    }
  }

  const loadQuickStats = async (): Promise<void> => {
    try {
      const response = await api.get('/stats/quick')
      const data = response.data

      lastLoginTime.value = data.last_login_time
      failedLoginsToday.value = data.failed_logins_today
      newUsersThisWeek.value = data.new_users_week
    } catch (error) {
      console.error('Failed to load quick stats:', error)
    }
  }

  return {
    userStats,
    sessionStats,
    orgStats,
    clientStats,
    activityData,
    loginDistribution,
    recentActivities,
    lastLoginTime,
    failedLoginsToday,
    newUsersThisWeek,
    loadUserStats,
    loadSessionStats,
    loadOrgStats,
    loadClientStats,
    loadActivityData,
    loadLoginDistribution,
    loadRecentActivities,
    loadQuickStats
  }
})

File: admin-service/src/handlers/stats.rs
use axum::{extract::State, http::StatusCode, response::Json, Extension};
use serde_json::{json, Value};
use tracing::info;
use std::sync::Arc;
use tokio::sync::RwLock;

use crate::{config::Config, jwt::JwtVerifier, storage::AdminStorage, models::Claims};

type AppState = (Arc<RwLock<AdminStorage>>, Arc<JwtVerifier>, Config);

pub async fn users_stats(
    State((storage, _, _)): State<AppState>,
    Extension(claims): Extension<Claims>,
) -> Result<Json<Value>, StatusCode> {
    info!(
        service = "admin-service",
        event = "users_stats_request",
        requested_by = %claims.sub
    );

    let storage_guard = storage.read().await;
    let users_count = storage_guard.users_count();

    // Mock data for demo - in real implementation, calculate from actual data
    let response = json!({
        "total_users": users_count,
        "growth_percentage": 12.5,
        "active_users": users_count - 1,
        "new_users_today": 2,
        "new_users_week": 8
    });

    Ok(Json(response))
}

pub async fn sessions_stats(
    State((storage, _, _)): State<AppState>,
    Extension(claims): Extension<Claims>,
) -> Result<Json<Value>, StatusCode> {
    info!(
        service = "admin-service",
        event = "sessions_stats_request",
        requested_by = %claims.sub
    );

    // Mock data for demo
    let response = json!({
        "active_sessions": 15,
        "change_percentage": 8.3,
        "total_sessions_today": 42,
        "unique_users_today": 28
    });

    Ok(Json(response))
}

pub async fn organizations_stats(
    State((storage, _, _)): State<AppState>,
    Extension(claims): Extension<Claims>,
) -> Result<Json<Value>, StatusCode> {
    info!(
        service = "admin-service",
        event = "organizations_stats_request",
        requested_by = %claims.sub
    );

    let storage_guard = storage.read().await;
    let orgs_count = storage_guard.organizations_count();

    let response = json!({
        "total_organizations": orgs_count,
        "growth_percentage": 5.2,
        "active_organizations": orgs_count
    });

    Ok(Json(response))
}

pub async fn clients_stats(
    State((storage, _, _)): State<AppState>,
    Extension(claims): Extension<Claims>,
) -> Result<Json<Value>, StatusCode> {
    info!(
        service = "admin-service",
        event = "clients_stats_request",
        requested_by = %claims.sub
    );

    let storage_guard = storage.read().await;
    let clients_count = storage_guard.clients_count();

    let response = json!({
        "total_clients": clients_count,
        "change_percentage": -2.1,
        "active_clients": clients_count
    });

    Ok(Json(response))
}

pub async fn activity_data(
    State((storage, _, _)): State<AppState>,
    Extension(claims): Extension<Claims>,
) -> Result<Json<Value>, StatusCode> {
    info!(
        service = "admin-service",
        event = "activity_data_request",
        requested_by = %claims.sub
    );

    // Mock activity data for the last 7 days
    let response = json!([
        {
            "date": "2024-11-01",
            "logins": 45,
            "registrations": 3,
            "active_users": 42
        },
        {
            "date": "2024-11-02",
            "logins": 52,
            "registrations": 5,
            "active_users": 48
        },
        {
            "date": "2024-11-03",
            "logins": 38,
            "registrations": 2,
            "active_users": 35
        }
    ]);

    Ok(Json(response))
}

pub async fn login_distribution(
    State((storage, _, _)): State<AppState>,
    Extension(claims): Extension<Claims>,
) -> Result<Json<Value>, StatusCode> {
    info!(
        service = "admin-service",
        event = "login_distribution_request",
        requested_by = %claims.sub
    );

    // Mock login distribution by organization
    let response = json!([
        {
            "org": "default",
            "count": 125,
            "percentage": 45.5
        },
        {
            "org": "group-7a",
            "count": 89,
            "percentage": 32.4
        },
        {
            "org": "group-8b",
            "count": 61,
            "percentage": 22.1
        }
    ]);

    Ok(Json(response))
}

pub async fn recent_activities(
    State((storage, _, _)): State<AppState>,
    Extension(claims): Extension<Claims>,
) -> Result<Json<Value>, StatusCode> {
    info!(
        service = "admin-service",
        event = "recent_activities_request",
        requested_by = %claims.sub
    );

    // Mock recent activities
    let response = json!([
        {
            "id": "act-001",
            "type": "login",
            "user_email": "user@example.com",
            "org": "default",
            "timestamp": "2024-11-03T15:30:00Z",
            "details": "Successful login from 192.168.1.100"
        },
        {
            "id": "act-002",
            "type": "registration",
            "user_email": "newuser@example.com",
            "org": "group-7a",
            "timestamp": "2024-11-03T14:45:00Z",
            "details": "New user registration"
        }
    ]);

    Ok(Json(response))
}

pub async fn quick_stats(
    State((storage, _, _)): State<AppState>,
    Extension(claims): Extension<Claims>,
) -> Result<Json<Value>, StatusCode> {
    info!(
        service = "admin-service",
        event = "quick_stats_request",
        requested_by = %claims.sub
    );

    let response = json!({
        "last_login_time": "2024-11-03T15:30:00Z",
        "failed_logins_today": 3,
        "new_users_week": 12
    });

    Ok(Json(response))
}